Importing the Libraries

In [1]:
import numpy as np
import pandas as pd

Dataset from kaggle - SPAM SMS DETECTION  
(https://www.kaggle.com/datasets/uciml/sms-spam-collection-dataset?select=spam.csv)

It is the dataset which have some (5572 rows) data of the sms.  
This data shows that, whether a sms is spam or not.

Reading the dataset

In [2]:
df = pd.read_csv("C:/Users/Hp/Downloads/codsoft intership/spam.csv", encoding="latin-1")
df

v1                                                 v2 Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  
1           NaN        NaN  
2           NaN        NaN  
3           NaN        NaN  
4           NaN        NaN  
...         ...        ...  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN  

[5572 rows x 5 columns]

The last 3 columns are not valuable, so will drop it.  
And rename the column 1,2 as :  
v1 => class  
v2 => text

In [3]:
df = df.drop(["Unnamed: 2", "Unnamed: 3", "Unnamed: 4"], axis=1)
df = df.rename(columns={"v1": "Class", "v2": "text"})
df

Class                                               text
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham              Will Ì_ b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

Class => shows ham or spam.  
text => the sms/message. 

In [4]:
df["Class"].unique()

array(['ham', 'spam'], dtype=object)

Will change the "Class" to Binary (0/1).  
ham => 0  
spam => 1

In [5]:
df["Class"] = df.Class.map({"ham": 0, "spam": 1})
df

Class                                               text
0         0  Go until jurong point, crazy.. Available only ...
1         0                      Ok lar... Joking wif u oni...
2         1  Free entry in 2 a wkly comp to win FA Cup fina...
3         0  U dun say so early hor... U c already then say...
4         0  Nah I don't think he goes to usf, he lives aro...
...     ...                                                ...
5567      1  This is the 2nd time we have tried 2 contact u...
5568      0              Will Ì_ b going to esplanade fr home?
5569      0  Pity, * was in mood for that. So...any other s...
5570      0  The guy did some bitching but I acted like i'd...
5571      0                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [6]:
df.isnull().sum()

Class    0
text     0
dtype: int64

In [7]:
df.duplicated().sum()

403

There are duplicates.  
So will be removing duplicates.

In [8]:
df = df.drop_duplicates(keep="first")
df

Class                                               text
0         0  Go until jurong point, crazy.. Available only ...
1         0                      Ok lar... Joking wif u oni...
2         1  Free entry in 2 a wkly comp to win FA Cup fina...
3         0  U dun say so early hor... U c already then say...
4         0  Nah I don't think he goes to usf, he lives aro...
...     ...                                                ...
5567      1  This is the 2nd time we have tried 2 contact u...
5568      0              Will Ì_ b going to esplanade fr home?
5569      0  Pity, * was in mood for that. So...any other s...
5570      0  The guy did some bitching but I acted like i'd...
5571      0                         Rofl. Its true to its name

[5169 rows x 2 columns]

In [9]:
print(df["Class"].value_counts())
pd.crosstab(index= df["Class"], columns="count", normalize=True)

Class
0    4516
1     653
Name: count, dtype: int64


col_0    count
Class         
0      0.87367
1      0.12633

Spam message/text is 653, i.e; 12.63 % of total text/message.

Now, again importing the Libraries to perform a model on dataset.

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer

#### Logistic Regression

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Spliting the dataset into "Independent" and "Dependent" variables.

In [12]:
x = df.iloc[:,1]
x

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                Will Ì_ b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: text, Length: 5169, dtype: object

In [13]:
y = df.iloc[:,0]
y

0       0
1       0
2       1
3       0
4       0
       ..
5567    1
5568    0
5569    0
5570    0
5571    0
Name: Class, Length: 5169, dtype: int64

In [14]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=42)

Converting text data into numerical value 

In [15]:
feature_extract = TfidfVectorizer(min_df=1, stop_words='english', lowercase=True)

x_train_features = feature_extract.fit_transform(x_train)
x_test_features = feature_extract.transform(x_test)

applying logistic regression model.

In [16]:
log_model = LogisticRegression()
log_model.fit(x_train_features, y_train)

LogisticRegression()

In [17]:
logistic_train_data_pred = log_model.predict(x_train_features)
print("Accuracy of training data",accuracy_score(logistic_train_data_pred,y_train))

Accuracy of training data 0.9613059250302297


In [18]:
logistic_test_data_pred = log_model.predict(x_test_features)
print("Accuracy of testing data:",accuracy_score(logistic_test_data_pred,y_test))

Accuracy of testing data: 0.9642166344294004


#### SVM

In [19]:
from sklearn.svm import SVC

Applying SVM model, to check whether gettting more accuracy score from previous. 

In [20]:
svm_model = SVC()
svm_model.fit(x_train_features,y_train)

SVC()

In [21]:
svm_train_data_pred = svm_model.predict(x_train_features)
print("Accuracy of training data:",accuracy_score(svm_train_data_pred,y_train))

Accuracy of training data: 0.9983071342200726


In [22]:
svm_test_data_pred = svm_model.predict(x_test_features)
print("Accuracy of testing data:",accuracy_score(svm_test_data_pred,y_test))

Accuracy of testing data: 0.9787234042553191


As per the output, accuracy of "testing data" is more good by "SVM model".  
So, "SVM model" is performing better than "Logistic Regressiion model". 

In [23]:
input_sms = ["Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's"]
print(input_sms)

["Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's"]


In [24]:
input_sms_feature = feature_extract.transform(input_sms)
predict = svm_model.predict(input_sms_feature)
print(predict)

if predict[0] == 1:
    print("Spam")
else:
    print("Ham")

[1]
Spam
